File: python/array_and_linkedlist/list.py
"""
列表及其相关操作，可以将列表看为动态数组
1. 随机访问 O(1)
2. 更新指定索引元素 O(1)
3. 插入元素 O(n)
4. 追加元素 O(1)
5. 删除元素 O(n)
6. 遍历 O(n)
7. 合并列表 O(m+n)
8. 排序 O(n)
"""


def access(nums: list[int], index: int) -> int:
    """
    访问元素，O(1)
    :param nums: 数据源
    :param index: 索引
    :return: int
    """
    return nums[index]


def update(nums: list[int], index: int, val: int) -> None:
    """
    更新元素，O(1)
    :param nums: 数据源
    :param index: 索引
    :param val: 更新的值
    :return:
    """
    nums[index] = val


def insert(nums: list[int], index: int, val: int) -> None:
    """
     插入元素，O(n)
    :param nums:
    :param index:
    :param val:
    :return:
    """
    nums.insert(index, val)


def append(nums: list[int], val: int) -> None:
    """
    追加元素，O(1)
    :param nums:
    :param val:
    :return:
    """
    nums.append(val)


def pop(nums: list[int], index: int) -> None:
    """
    删除元素，O(n)
    :param nums:
    :param index:
    :return:
    """
    nums.pop(index)


def reverse(nums: list[int]) -> None:
    """
    遍历列表，O(n)
    :param nums:
    :return:
    """
    for num in nums:
        print(num)


def extend(nums: list[int], nums1: list[int]) -> list[int]:
    """
    拼接列表, O(m+n)
    :param nums:
    :param nums1:
    :return:
    """
    return nums + nums1


def sort(nums: list[int]) -> None:
    """
    排序， O(n)
    :param nums:
    :return:
    """
    nums.sort()


def main():
    nums1: list[int] = []
    nums: list[int] = [1, 3, 2, 5, 4]

    # 访问元素
    assert access(nums, 0) == 1
    # 更新元素
    update(nums, 0, 10)
    assert nums[0] == 10

    # 插入
    insert(nums, 1, 200)
    assert nums == [10, 200, 3, 2, 5, 4]

    # 追加
    append(nums, 300)
    assert nums == [10, 200, 3, 2, 5, 4, 300]

    # 删除
    pop(nums, 3)
    assert nums == [10, 200, 3, 5, 4, 300]

    # 遍历
    print("[reverse nums]")
    reverse(nums)

    # 合并两个列表
    append(nums1, 1000)
    nums2 = extend(nums, nums1)
    assert nums2 == [10, 200, 3, 5, 4, 300, 1000]

    # 排序
    sort(nums)
    assert nums == [3, 4, 5, 10, 200, 300]


if __name__ == "__main__":
    main()

File: go/array_and_linkedlist/list.go
// 列表及其相关操作，基于 slice 实现
package main

import (
    "fmt"
    "sort"
)

func main() {
    // 未初始化和初始化的列表
    // nums1 := []int{}
    nums := []int{1, 3, 2, 5, 4}

    // 访问元素 access O(1)
    fmt.Printf("[access 2] %d\n", nums[2])

    // 插入元素 insert O(n)
    // 索引 2 插入数字 10
    nums = append(nums[:2], append([]int{6}, nums[2:]...)...)
    fmt.Printf("[append index: 2 value: 6] %v\n", nums)

    // 追加元素 append O(1) 10
    nums = append(nums, 10)
    fmt.Printf("[append -1] %d\n", nums[6])

    // 根据索引删除元素 O(n) 2
    nums = append(nums[:2], nums[3:]...)
    fmt.Printf("[pop] %v\n", nums)

    // 遍历列表 reverse O(n)
    fmt.Print("[reverse]")
    for _, num := range nums {
        fmt.Printf("%d ", num)
    }
    fmt.Print("\n")

    // 拼接元素 extend O(n)
    nums1 := []int{6, 7}
    nums = append(nums, nums1...)
    fmt.Printf("[extend] %v\n", nums)

    // 排序
    sort.Ints(nums)
    fmt.Printf("[sort] %v\n", nums)

}
